File: solutions/scala/spiral-matrix/src/main/scala/SpiralMatrix.scala
object SpiralMatrix:
  def spiralMatrix(size: Int, start: Int = 1): List[List[Int]] = size match
    case 0 => Nil
    case 1 => List(List(start))
    case _ =>
      val step = size - 1
      val innerMatrix = spiralMatrix(size - 2, start + step * 4)
      val middleRows = 1 until step map { i =>
        val left = start + step * 4 - i
        val right = start + step + i
        left :: innerMatrix(i - 1) ::: right :: Nil
      }
      val topRow = start to start + step
      val bottomRow = start + step * 3 to start + step * 2 by -1
      topRow.toList :: middleRows.toList ::: bottomRow.toList :: Nil

File: solutions/python/spiral-matrix/spiral_matrix.py
def spiral_matrix(size: int) -> list[list[int]]:
    """Creates a square matrix of given size filled with numbers in spiral order.

    This function generates a square matrix where numbers are arranged in a clockwise
    spiral pattern starting from 1 in the top-left corner.

    Args:
        size: An integer representing both the number of rows and columns
            in the square matrix.

    Returns:
        list[list[int]]: A square matrix of size x size containing integers
            arranged in clockwise spiral order starting from 1.

    Example:
        >>> spiral_matrix(3)
        [[1, 2, 3],
         [8, 9, 4],
         [7, 6, 5]]
    """

    def construct_matrix(rows: int, columns: int, start: int) -> list[list[int]]:
        if rows == 0:
            return []
        top_row = list(range(start, start + columns))
        inner_matrix = reversed(construct_matrix(columns, rows - 1, start + columns))
        return [top_row] + [list(row) for row in zip(*inner_matrix)]

    return construct_matrix(size, size, 1)
